### PR TITLE
[feat] #15 Add S3 storage strategy 

### PR DESCRIPTION
**Pull request by choinek**
* Sorry for a few unnecessary changes in tabs, spaces, or lines - I don't usually code in Python and currently don't have an IDE for it.

I added support for s3 storage - it handles save, list, load and delete.

I was considering the best place for tokens, and I believe environment variables are the most suitable option.

While adding instructions to the README, I slightly modified the structure of the 'storage profiles' section.

1) working ocr_upload example-mri.pdf -> markdown -> s3
```
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py ocr_upload --file examples/example-mri.pdf --ocr_cache --prompt_file=examples/example-mri-remove-pii.txt --storage_profile=s3
Namespace(command='ocr_upload', file='examples/example-mri.pdf', ocr_cache=True, prompt=None, prompt_file='examples/example-mri-remove-pii.txt', model='llama3.1', strategy='marker', print_progress=True, storage_profile='s3', storage_filename=None)
File uploaded successfully. 
(...)
```
2) working ocr_upload example-invoice.pdf -> markdown -> s3
```(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py ocr_upload --file examples/example-invoice.pdf --prompt_file examples/example-invoice-remove-pii.txt --storage_profile s3 
Namespace(command='ocr_upload', file='examples/example-invoice.pdf', ocr_cache=True, prompt=None, prompt_file='examples/example-invoice-remove-pii.txt', model='llama3.1', strategy='marker', print_progress=True, storage_profile='s3', storage_filename=None)
File uploaded successfully.
(...)
```

3) Check if the files were really uploaded.
image

3) Working s3 list_files
```
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py list_files --storage_profile=s3

example-invoice.md
example-mri.md
```

4) Working s3 load_file
```
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py load_file --file_name="example-invoice.md" --storage_profile=s3

Here is the corrected text with personal information removed and spelling issues fixed:

**Acme Invoice Ltd**
(...)
```

5) Working s3 delete_file 
```
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py list_files --storage_profile=s3

example-invoice.md
example-mri.md
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py delete_file --file_name "example-mri.md" --storage_profile s3
File example-mri.md deleted successfully.
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py list_files --storage_profile=s3                                

example-invoice.md
```

6) working s3 storage_filename
```
(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py ocr_upload --file examples/example-mri.pdf --ocr_cache --prompt_file=examples/example-mri-remove-pii.txt --storage_filename "invoices/{Y}/{file_name}-{Y}-{mm}-{dd}.md" --storage_profile=s3
Namespace(command='ocr_upload', file='examples/example-mri.pdf', ocr_cache=True, prompt=None, prompt_file='examples/example-mri-remove-pii.txt', model='llama3.1', strategy='marker', print_progress=True, storage_profile='s3', storage_filename='invoices/{Y}/{file_name}-{Y}-{mm}-{dd}.md')
File uploaded successfully.

(.venv) ➜  pdf-extract-api git:(main) ✗ python client/cli.py list_files --storage_profile=s3

example-invoice.md
example-mri.md
**invoices/2024/example-mri-2024-11-16.md**

```
image

Also you can find the policy I used, which is working, in the README.
image

### CODE REVIEW

**Comment by choinek** [comment]
w8, I'm adding support for format_file_name

File: app/storage_strategies/aws_s3.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            's3',
+            aws_access_key_id=self.access_key,
+            aws_secret_access_key=self.secret_access_key,
+            region_name=self.region
+        )
+
+    def _resolve_placeholder(self, value):
+        if not value:
+            return None
+        return Template(value).substitute(os.environ)
```
Comment (pkarw) [inline app/storage_strategies/aws_s3.py:26]: Nice one! We'll need to apply it in the other storage strategies as well, great idea 🙏

File: app/storage_strategies/aws_s3.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        formatted_file_name = self.format_file_name(file_name, dest_file_name)
+
+        try:
+            self.s3_client.put_object(
+                Bucket=self.bucket_name,
+                Key=formatted_file_name,
+                Body=content.encode('utf-8')
+            )
+            print(f"File {formatted_file_name} saved to S3 bucket {self.bucket_name}.")
+        except NoCredentialsError:
```
Comment (pkarw) [inline app/storage_strategies/aws_s3.py:38]: In the other strategies errors are not handled on this level - raised up and handled in the storage_manager - API

I think it should be the way (not handling exception here but leaving it to level up) to make sure they are always resolved the same way

Wdyt?

**Review by pkarw — CHANGES_REQUESTED** [review]
The only thing to change is to pass on handling exceptions bc they’re handled level up for example by Celery task processing lib

File: app/storage_strategies/aws_s3.py (hunk 2 above)
Comment (choinek) [reply app/storage_strategies/aws_s3.py:38]: I think we should handle exceptions specific to the source libraries used in the storage_strategies and re-throw them as a common exception to ensure consistent error handling. Sometimes, we may need to add specific information to an exception, and this should be done at the strategy level, not in the storage_manager. What do you think about this approach? If it's acceptable, I will re-throw the exceptions. If not, I will implement it as you suggest.

File: app/storage_strategies/aws_s3.py (hunk 1 above)
Comment (choinek) [reply app/storage_strategies/aws_s3.py:26]: Maybe we should move it to StorageStrategy?

File: app/storage_strategies/aws_s3.py (hunk 2 above)
Comment (pkarw) [reply app/storage_strategies/aws_s3.py:38]: Fair point. Please handle and re-throw

Comment (choinek) [reply app/storage_strategies/aws_s3.py]: Ale ta dyskusja spowodowała, że nieco zluzowałem z exceptionami, których i tak sensownie nie wzbogacę - przerzuciłem sprawdzenie połączenia na inita, a same metody rethrowują wzbogacony exception aby pomoc w rozpoznaniu problemu jak np.
image
image

File: app/storage_strategies/aws_s3.py (hunk 1 above)
Comment (choinek) [reply app/storage_strategies/aws_s3.py]: Okay, I moved it and even improved it, so now it properly handles missing keys.

`Exception: Environment variable 'AWS_S3_BUCKET_NAME' is missing. Please set it in your environment or configuration.
`

**Comment by choinek** [comment]
@pkarw już można sprawdzać, zweryfikowałem po zmianach działanie każdej z metod i sytuacje, gdzie user ma ustawione puste envy oraz nie ma envów - w obu przypadkach komunikat jasno sugeruje co musi zrobić.

**Review by pkarw — APPROVED** [review]

**Review by pkarw — APPROVED** [review]

**Pull request merged**